### PR TITLE
Implement a basic conversation wrapper on top of ChatGPT

### PR DESCRIPTION
**Pull request by xacrimon**
This PR implements conversation via a websocket to ChatGPT with message persistence/loading of previous conversations. Currently only soft text responses are sent but in the future, special control JSON messages will also be sent to notify the frontend that it should suggest running a specific command to the user.

Since I'm gone until Tuesday next week on bank holidays. Here are some ideas and a sketch for what we need to do to move forward:
1. Set up one or more prompt messages inserted into the `ai.Chat` structure in the beginning of a conversation that informs it of it's task. This includes informing it of what data we need to collect, i.e labels and hostnames and to come up with a command to execute.
2. On message received by the user, clone the `ai.Chat` structure and give it a system message to self analyze the current conversation and attempt to extract this info or ask the user for other required info.
3. If it responds back with required data, i.e a command and some targets, we need to send a special WS message of a different kind to the frontend to tell it to suggest the action to the user. (frontend hits execution API if confirmed, if not we allow the user to refine and suggest again when reasonable)

The second main idea I think we need to look at is providing it with some cluster information as this vastly improves it's reasoning. I think we should look into implementing a simplified version of the GPTListIndex from LlamaIndex (actual logic is pretty simple) at first to get it working but some more clever solution will be required for larger datasets of cluster info. The two promising methods involve [APIs](https://python.langchain.com/en/latest/use_cases/apis.html) and [indexed data stores](https://python.langchain.com/en/latest/modules/indexes.html) which would allow us to filter out and provide the model with the needed context in the form of system messages in the `chat.AI` structure at conversation start.

### CODE REVIEW

File: api/proto/teleport/legacy/types/types.proto [hunk 1]
```diff
@@ -1671,6 +1671,14 @@ message AuthPreferenceSpecV2 {
     (gogoproto.jsontag) = "allow_headless,omitempty",
     (gogoproto.customtype) = "BoolOption"
   ];
+
+  // Assist is a set of options related to the Teleport Assist feature.
+  AssistOptions Assist = 16 [(gogoproto.jsontag) = "assist,omitempty"];
```
Comment (xacrimon) [inline api/proto/teleport/legacy/types/types.proto:1676]: This should go somewhere else, but I'm not sure where?

File: lib/ai/chat.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	return Message{
+		Role:    role,
+		Content: content,
+		Idx:     len(chat.messages) - 1,
+	}
+}
+
+func (chat *Chat) Complete(ctx context.Context, maxTokens int) (Message, error) {
+	request := openai.ChatCompletionRequest{
+		Model:     openai.GPT3Dot5Turbo,
```
Comment (jakule) [inline lib/ai/chat.go:52]: Shouldn't we set it to GPT-4?

File: lib/ai/chat.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
+/*
```
Comment (xacrimon) [inline lib/ai/chat.go:1]: This file is in a new `ai` module to give space for related code for knowledge management and other things needed to make context-aware reasoning work with cluster data.

**Review by jakule — APPROVED** [review]
LGTM. It'd be nice to add some comments, but the code is clear enough for me that I'm fine with merging it in the current form.

File: api/proto/teleport/legacy/types/types.proto [hunk 4]
```diff
@@ -1671,6 +1671,14 @@ message AuthPreferenceSpecV2 {
...
+}
+
+// AssistOptions is a set of options related to the Teleport Assist feature.
+message AssistOptions {
+  string ApiKey = 1 [(gogoproto.jsontag) = "api_key,omitempty"];
```
Comment (zmb3) [inline api/proto/teleport/legacy/types/types.proto:1681]:
```suggestion
  string APIKey = 1 [(gogoproto.jsontag) = "api_key,omitempty"];
```

Match conventions throughout the rest of this file.

File: lib/config/fileconf.go [hunk 5]
```diff
@@ -1337,6 +1349,18 @@ func (dt *DeviceTrust) Parse() (*types.DeviceTrust, error) {
 	}, nil
 }
 
+// AssistOptions is a set of options related to the Teleport Assist feature.
+type AssistOptions struct {
+	// APIKey is the API key used to authenticate requests to the Teleport Assist API.
+	ApiKey string `yaml:"api_key,omitempty"`
```
Comment (zmb3) [inline lib/config/fileconf.go:1355]:
```suggestion
	APIKey string `yaml:"api_key,omitempty"`
```

File: lib/web/assistant.go [hunk 6]
```diff
@@ -28,16 +29,35 @@ import (
...
 func (h *Handler) assistant(w http.ResponseWriter, r *http.Request, _ httprouter.Params, sctx *SessionContext) (any, error) {
-	authClient, err := sctx.GetClient()
+	// moved into a seperate function for error management/debug purposes
+	err := runAssistant(h, w, r, sctx)
 	if err != nil {
+		h.log.Warn(trace.DebugReport(err))
 		return nil, trace.Wrap(err)
 	}
 
+	return nil, nil
```
Comment (zmb3) [inline lib/web/assistant.go:51]:
```suggestion
	return Ok(), nil
```

I think?

**Review by ryanclark — APPROVED** [review]

File: api/proto/teleport/legacy/types/types.proto (hunk 4 above)
Comment (jakule) [reply api/proto/teleport/legacy/types/types.proto:1681]: Changed

**Pull request merged**